### ISSUE TITLE
Triggering on type formatting on deletions

### ISSUE THREAD

**Issue by vinistock** [microsoft/vscode-languageserver-node#1187]
Generally, on type formatting is used to autocomplete code as the developer types specific characters. However, there are also use cases for doing similar operations on deletions.

For example, when doing string interpolation in Ruby, we are currently closing brace automatically using on type formatting (from `"#{"` to `"#{}"`), but we'd also be interested in removing that closing brace in case the user deletes the opening one with backspace (from `"#{}"` to `"#"`). There are other use cases involving `end` tokens and pipes too.

Is it possible to trigger on type formatting on backspace? And if so, what's the trigger character?

**Comment by dbaeumer** [comment]
I think using on type formatting for this is not a good idea since it is not designed for this. 

@aeschli I remember you have some comparable adjustments in HTML for tags. How do you do this there?

**Comment by aeschli** [comment]
Yes, we have a custom LSP request for adding quotes and for closing tags.
We generalized it by calling it `autoInsertion`. Given a trigger character the autoInsertion provider is asked for an edit.

**Comment by vinistock** [comment]
For HTML tags, you could do something similar with deletions. For example, if you have `<p></p>` and you completely remove the closing tag, the opening tag can be automatically removed (same for the other way around).

Is there interest in having this for language servers? In my head, it sounds like a part of on type formatting, but Dirk mentioned that's not adequate at the moment. Would it be a new request or a modification to on type formatting?

**Comment by dbaeumer** [comment]
If we add this is should be a new request.

**Comment by vinistock** [comment]
I did a quick attempt to implement this as a custom request, but I now realize the issue you mentioned. I tried hooking it up on `DidChangeTextDocument`, but for deletions the `contentChanges` are simply empty. We don't know which key press triggered it or what is the location of the edit.

Are there any alternatives way to implement this? Feels like it requires some changes in VS Code itself to support it.

**Comment by dbaeumer** [comment]
@aeschli any additional tips
